Defer queue creation / activate proxy queue

faec opened the pull request:
(This is probably ready for comments but I've left it as a draft while I work on testing.)

Defer creation of the beats queue until the full output configuration is received; Activate the proxy queue when the Shipper output is in use.

The proxy queue is meant to track event acknowledgments without storing event data in memory, to avoid a doubled memory cost for events stored in both Beats and the Shipper. However, [activating it](https://github.com/elastic/beats/issues/34396) has required multiple refactoring passes, since previously the queue was created globally before receiving any output configuration. Now that previous work has moved queue logic into `outputController`, this PR completes the work by activating the queue-related hooks so queue creation is delayed and queue clients are blocked until the output is active.

The important changes are:
- `outputs.Group` now has a new field, `QueueSettings`, with which an output can specify the settings to use when creating the queue.
- `outputController.Set`, which gives the output controller its output worker group, is now responsible for creating the queue when a nonempty output is assigned (previously this was done on initialization).
- `outputController.queueProducer` can now be called before the queue is created, in which case it accumulates requests that then block until queue creation.
- refactor: the queue's acknowledgment callback is now a parameter to queue creation rather than an element of the various `Settings` structs. The callback is used for bookkeeping in the pipeline and needs to be controlled by the queue creator, while everything else in `{memqueue,diskqueue,proxyqueue}.Settings` can be safely set by the output.
- refactor: pipeline waitgroup / event tracking. Various event transitions involve calling into the pipeline observer or the pipeline's global event wait group. These are now passed down as creation parameters to the `client` and `outputController` so both components can update them in a consistent way instead of distributing the logic across various callbacks and explicit `Pipeline` pointers.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] ~~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## How to test this PR locally

There are two paths to testing this PR: one is to make sure that normal queue settings continue to apply, both standalone and under agent. This should just involve making sure that events reach the output without errors. The other is to test with the shipper output and verify that data can flow to the shipper, that event acknowledgments are received, and that the beat memory use remains low even when many events are queued by the shipper.

## Related issues

- Resolves https://github.com/elastic/beats/issues/34396

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-35118/detail/PR-35118/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-35118/detail/PR-35118/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-35118/detail/PR-35118/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-35118/detail/PR-35118/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2023-04-18T20:00:47.570Z&rangeTo=2023-04-18T20:20:47.570Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=ac101228b1f088a3cad440155c6701de&transactionId=d1ded59896e1e1a8)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-04-18T20:10:47.570+0000


* Duration: 64 min 14 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 4842  |
| Skipped      | 289 |
| Total        | 5131   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![7](https://img.shields.io/badge/7%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-35118/detail/PR-35118/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-35118/runs/1/steps/3371/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'windows-2022'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-35118/runs/1/steps/3419/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'windows-2022'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-35118/runs/1/steps/3429/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'windows-2016'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-35118/runs/1/steps/3483/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'windows-2022'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-35118/runs/1/steps/3491/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'windows-2022'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-35118/runs/1/steps/3502/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'windows-2022'.[0m</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 1 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-35118/runs/1/steps/4414/log/?start=0">here</a></li>
<li>Description: <code>untar: step failed with error null</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

faec commented:
I've fixed the most relevant tests and added unit tests for the new behavior. I'm still investigating some CI failures that I can't yet reproduce locally but I think it's ready for review.

fearful-symmetry commented:
Oddly enough, looking at the log files generated by the beat, I can see the log line that the test is looking for:
```
 cat /home/alexk/go/src/github.com/elastic/beats/libbeat/build/system-tests/run/test_seccomp.Test.test_seccomp_installed199/mockbeat-20230420.ndjson | grep -i syscall
{"log.level":"debug","@timestamp":"2023-04-20T14:46:48.989-0700","log.logger":"seccomp","log.origin":{"file.name":"seccomp/seccomp.go","file.line":117},"message":"Loading syscall filter","service.name":"mockbeat","seccomp_filter":{"no_new_privs":true,"flag":"tsync","policy":
{"log.level":"info","@timestamp":"2023-04-20T14:46:48.989-0700","log.logger":"seccomp","log.origin":{"file.name":"seccomp/seccomp.go","file.line":124},"message":"Syscall filter successfully installed","service.name":"mockbeat","ecs.version":"1.6.0"}
```

fearful-symmetry commented on libbeat/publisher/pipeline/pipeline.go:
```diff
@@ -386,3 +389,26 @@ func (p *Pipeline) createEventProcessing(cfg beat.ProcessingConfig, noPublish bo
+// parsing errors: if there is an error in the queue config, we want it to
+// show up as fatal during initialization, even if the queue itself isn't
+// created until later.
+func queueSettingsForUserConfig(queueType string, userConfig *conf.C, inQueueSize int) (interface{}, error) {
```
That bare interface really bugs me, even if I understand why we ended up with it. At very least I kinda wonder if we could have some kind of named interface or something, just for the sake of making the API itself look less opaque.

faec replied:
It bugs me too and it wasn't my first choice. The constraint that led me here is that we need to be able to fail during initialization if we get an invalid queue config up front, therefore we need to call `SettingsFromConfig` for the appropriate queue since they all have incompatible settings, but we also need that to be overridable by something the output can create.
- Alternative: We could panic later when we realize the config is invalid. I don't like this one.
- Alternative: We could report an error when we see it but fall back on the default memory queue. (I'm open to this one but in the end I decided quitting on initialization was preferable.)

The settings alone aren't enough to create the queue, we also need the logger (which we could get from other sources in a pinch) and the queue's global acknowledgment callback (which we can't) so we can't have the output actually _create_ the queue. On the other hand, just accepting a queue type / `*conf.C` is also awkward: the proxy queue isn't user-configurable, so a natural future choice like passing along a `queue` subtree from an `output` configuration would expose that internal type. (Besides which, to me an `interface{}` that is a placeholder for 3 concrete types is still a lot easier to work with than a `*conf.C` that could contain literally anything).

The old way of handling this tension was to use factories, which gives thing a nice name but makes the internals too opaque: a factory that just took the callback and created a queue would work, but there would be no way to know what type of queue we'd created, which we currently report.

Which... having talked myself into a corner, maybe we just need to add a `QueueType()` to the queue interface, and have both the pipeline initialization and the outputs give a queue factory for this field? At this point it could be a `func(*logp.Logger, func(eventCount int)) (queue.Queue, error)` which is more consistent than the old types, and we could still fall back on the memory queue on error (knowing that an honest queue creation can only fail for the disk queue). I dislike losing access to the settings objects but right now we don't actually need them so this might make sense, thoughts?

faec replied:
Ok, I switched to a version using factories, see if you feel better about this one ^^

fearful-symmetry commented:
So, the tests produce a huge number of these log lines:
```
{"log.level":"error","@timestamp":"2023-04-20T15:19:39.507-0700","log.logger":"publish","log.origin":{"file.name":"pipeline/controller.go","file.line":224},"message":"outputController received new output configuration when queue is already active","ecs.version":"1.6.0"}
```

Not sure how much of that is the tests, but it makes me worried we might produce a bit of log spam with it.

leehinman reviewed (changes requested):
Also curious about the large number of log messages that Alex saw.

faec commented:
> So, the tests produce a huge number of these log lines:
> 
> ```
> {"log.level":"error","@timestamp":"2023-04-20T15:19:39.507-0700","log.logger":"publish","log.origin":{"file.name":"pipeline/controller.go","file.line":224},"message":"outputController received new output configuration when queue is already active","ecs.version":"1.6.0"}
> ```
> 
> Not sure how much of that is the tests, but it makes me worried we might produce a bit of log spam with it.

I know why that is, and I felt conflicted about how to handle it, but the good news is that it won't happen in production: the cause is `TestOutputReload` in `test_controller.go`, which manually reloads the pipeline ~5000 times while simultaneously sending events through it to make sure nothing is lost.

I'm not sure the exact motivation, considering a real pipeline is only ever reloaded once. What it de facto tests is that the target of an `eventConsumer` can be changed during live event flow without dropping anything, and that's nice to know even though we don't rely on it, so I was reluctant to delete the test outright, but in a live beat it would be considered an error condition (and it would be extremely unusual for it to happen more than once, so I'm not worried about error spam).

So I'm not sure how to resolve the tension there -- the test is somewhat useful to keep currently-working asynchronous code from degrading, but it really is an error condition in current code so it necessarily spams the pipeline logger during tests. I could give the pipeline a nil logger for that test, but that would prevent logging of things we _do_ want to see. Maybe it's still worth it considering the noise, and just leave an explanatory comment saying to reactivate the logger if the test ever fails? Open to suggestions :-)

faec commented on libbeat/publisher/pipeline/controller.go:
```diff
@@ -169,45 +179,99 @@ func (c *outputController) Reload(
-	return c.queue.Producer(config)
+	c.queueLock.Lock()
+	if c.queue != nil {
+		defer c.queueLock.Unlock()
```
Not in this case (I've added a more detailed comment, sorry I meant to include one already) -- because in the other branch, where the queue doesn't exist, we need to manually unlock so we don't hold the mutex while waiting for the queue to be created.

faec commented:
> Oddly enough, looking at the log files generated by the beat, I can see the log line that the test is looking for:
> ...

Yes indeed, my local tests confirm that the exact same log line (modulo timestamp) appears in both cases but only my PR branch considers it a failure. So it must be something in how the testing framework scans logs? Very mysterious, especially since it doesn't seem to affect any other tests. Hopefully getting close to pinning down the problem...

fearful-symmetry commented on libbeat/publisher/queue/queue.go:
```diff
@@ -72,6 +74,8 @@ type Queue interface {
 	Metrics() (Metrics, error)
 }
 
+type QueueFactory func(logger *logp.Logger, ack func(eventCount int)) (Queue, error)
```
Small nit: can we get a comment on the QueueFactory type? Other than that, LGTM.

faec commented:
Found the full issue with the failing test and it's somewhat ugly:

Beat shutdown blocks on the main `Beater.Run` routine returning. The beater is sent a `Stop` callback when there's a shutdown signal (ctrl-c, kill, etc), but until its main `Run` method returns the rest of the shutdown never happens. This is mostly fine _except_ in cases where no output is ever set, which currently seems like it can only happen when the beat is passed `-N` on the command line (this sets `publishDisabled`, which discards any output configuration before it is applied). In this case, beats that blocked on an initial `(*Pipeline).Connect` call will never unblock, because the pipeline doesn't know it's shutting down and therefore will let the blocked requests for a `queue.Producer` sit there forever.

Complicating this is the fact that `(*Pipeline).Close` itself is seemingly never called at all in most cases, even when shutdown does proceed properly (otherwise I could refactor it to be triggered by the beater shutdown callback -- but as things stand I'm wary of adding entirely new code paths to the shutdown process without more thorough testing). This drives home something we already knew, which is that we should fix the Beats shutdown procedure, since it's known to be skipping important steps and at this point that's interfering with other pipeline features.

My proposal for a short-term fix, until we have time for a more thorough cleanup of pipeline shutdown: have `queueProducer` check the `publishDisabled` flag and, if set, return a no-op producer, since we know that otherwise any attempt to connect to the pipeline will block forever (unless the beater is _extraordinarily_ responsible about how it handles shutdown signals, which doesn't seem to be the case with current beats). This solution feels not great to me, but it doesn't feel particularly worse than the preexisting special cases around `publishDisabled`. Thoughts / objections?

fearful-symmetry commented:
Didn't even know `-N` was a thing. Considering the flag explicitly disables the output, I think that returning a no-op publisher is fairly reasonable?
